File: content/snippets/go/design_patterns/lazy_initialization.go
package main

import (
	"fmt"
	"sync"
)

type Lazy struct{ Value int }

var (
	inst *Lazy
	once sync.Once
)

func Instance() *Lazy { once.Do(func() { inst = &Lazy{} }); return inst }
func main() {
	Instance().Value = 8
	fmt.Println(Instance().Value)
}

File: content/snippets/cpp/design_patterns/lazy_initialization.cpp
#include <memory>
#include <iostream>

class LazySingleton {
public:
    static LazySingleton& instance(){
        static LazySingleton inst;
        return inst;
    }
    int value = 0;
};
int main(){
    LazySingleton::instance().value = 10;
    std::cout<<LazySingleton::instance().value<<"\n";
}
